snowfall/objectives/mmi.py

```diff
@@ -32,20 +32,65 @@ def forward(
             texts: List,
             supervision_segments: torch.Tensor
     ) -> Tuple[torch.Tensor, int, int]:
-        num_graphs, den_graphs = self.graph_compiler.compile(texts, self.P)
+        num_graphs, den_graphs = self.graph_compiler.compile(
+            texts, self.P, replicate_den=False)
+
         dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
 
-        num_lats = k2.intersect_dense(num_graphs, dense_fsa_vec, 10.0)
-        den_lats = k2.intersect_dense(den_graphs, dense_fsa_vec, 10.0)
+        device = num_graphs.device
+
+        num_fsas = num_graphs.shape[0]
+        assert dense_fsa_vec.dim0() == num_fsas
+
+        assert den_graphs.shape[0] == 1
+
+        # the aux_labels of num_graphs is k2.RaggedInt
+        # but it is torch.Tensor for den_graphs.
+        #
+        # The following converts den_graphs.aux_labels
+        # from torch.Tensor to k2.RaggedInt so that
+        # we can use k2.append() later
+        ragged_shape = k2.ragged.regular_ragged_shape(
+            dim0=den_graphs.aux_labels.numel(), dim1=1).to(device)
```

Comment:
Just for me to make sure I understand, dim0 is the number of arcs (weights on arcs) and dim1 is the number of graphs (“batch size”)?

Reply:
Let me give you a concrete example.

Suppose you have the following FSA,
```
0 1 1 0.2
1 2 -1 0.3
2
```
which has only two arcs and three states. Let's call the above Fsa `fsa`.
```
fsa.labels == torch.tensor([1, -1], dtype=torch.int32)
fsa.labels.shape[0] == fsa.arcs.num_elements()

fsa.scores == torch.tensor([0.2, 0.3], dtype=torch.float32)
fsa.scores.shape[0] == fsa.arcs.num_eleements()
```
Note that there are two arcs and each arc has a **SCALAR** attribute, `label`. So its `labels` is a 1-D tensor.

Each arc also has an attribute, `score`, which also happens to be a 1-D tensor.

We can assign an arbitrary attribute to the arcs of `fsa`. For instance, we can assign a **scalar**
to each arc. In this case, since there are 2 arcs and each arc has a scalar attribute, we use
a 1-D tensor to represent this attribute, i.e.,
```
fsa.attr1 = torch.tensor([10, 20]) 

# It doesn't have to be of dtype torch.int32.
# The ONLY constraint is that `attr_tensor.shape[0] == num_arcs`
```

Also, we can assign a 1-D **TENSOR** (with 10 elements) to each arc. In this case, the resulting attribute is a 2-D tensor, i.e.,

```
fsa.attr2 = torch.rand(2, 10)

# It has to satisfy fsa.attr2.shape[0] == num_arcs
# Otherwise, it will raise an exception.
```

**CAUTION**: If the attribute is a torch.Tensor, then it requires that each arc's attribute has the same number of elements.
In the above case,  the first arc's attr2 has 10 entries, so is the second arc.

You can assign an N-D tensor attribute to each arc, if you like. The above examples demostrate `N==0` and `N==1`.


**WHAT** if different arcs have a different number of elements for a given attribute? 
In this case, we cannot use torch.Tensor anymore.
Instead, we can use `k2.RaggedInt` (`k2.RaggedFloat` can also be used, though it is not implemented in k2).

For instance, we give a third attribute `attr3` to `fsa`. The first arc's attribute value is [10, 20, 30] and the second one is [100]. It can be represented in k2 as
```
fsa.attr3 = k2.RaggedInt('[ [10 20 30] [100] ]')
```

We have
```
fsa.attr3.dim0() == fsa.arcs.num_elements()
```

Sometimes it may happen that some arcs's attribute value is empty. The following is an example:

```
fsa.attr4 = k2.RaggedInt('[ [1 2 3 4 5] [] ]')

fsa.attr4.dim0() == fsa.arcs.num_elements()
```
The first arc's attribute value is `[1, 2, 3, 4, 5]`, while the second arc's attribute value is empty.

---

Back to the current pull-request.

Suppose `den_graphs.aux_labels` is `torch.tensor([1, -1])`, the code changes it to
```
den_graphs.aux_labels = k2.RaggedInt('[ [1] [-1] ]')
```

The return value of
```
ragged_shape = k2.ragged.regular_ragged_shape(
            dim0=den_graphs.aux_labels.numel(), dim1=1).to(device)
```

is a `k2.RaggedShape`.  `dim0` represents the outer dimension, which is the number of arcs in this case.
The resulting ragged shape looks like
```
[ [x] [x] ]
```

There are two `[]` inside, which is specified by `dim0`.

There is one `x` inside each `[]`, which is specified by `dim1`.

Hope the above comment helps.

---

We are planning to write tutorials for Ragged tensors in k2. Please just wait for a few days or weeks.

Reply:
Thank you, that's a very clear description. It's impressive how much work you guys put into supporting ragged shapes in many possible ways.